src/VisualStudio/Core/Def/Copilot/VisualStudioCopilotCodeAnalysisService.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Copilot;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Settings;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Copilot;
+
+[ExportWorkspaceService(typeof(ICopilotCodeAnalysisService)), Shared]
+internal sealed class VisualStudioCopilotCodeAnalysisService : ICopilotCodeAnalysisService
+{
+    private readonly CopilotDiagnosticUpdateSource _copilotDiagnosticUpdateSource;
+    private readonly IDiagnosticsRefresher _diagnosticsRefresher;
+    private readonly IGlobalOptionService _globalOptions;
+    private readonly ConditionalWeakTable<Document, ConcurrentDictionary<string, (ImmutableArray<Diagnostic> Diagnostics, bool IsCompleteResult)>> _diagnosticsCache = new();
+    private ISettingsManager? _settingsManager;
+    private bool _executingAnalysis;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public VisualStudioCopilotCodeAnalysisService(
+        CopilotDiagnosticUpdateSource copilotDiagnosticUpdateSource,
+        IDiagnosticsRefresher diagnosticsRefresher,
+        IGlobalOptionService globalOptions,
+        IThreadingContext threadingContext,
+        IVsService<SVsSettingsPersistenceManager, ISettingsManager> settingsManagerService)
+    {
+        _copilotDiagnosticUpdateSource = copilotDiagnosticUpdateSource;
+        _diagnosticsRefresher = diagnosticsRefresher;
+        _globalOptions = globalOptions;
+        InitializeAsync(settingsManagerService, threadingContext.DisposalToken).Forget();
```

Comment:
what is going on here?  

untracked async work means we can't test this at all.